### PR TITLE
Move forward changes

### PR DESCRIPTION
**Pull request by zdasaro**
I added moveForward(int distance) to the mouse API. This allows the mouse to move forward several tiles at once.

### CODE REVIEW

File: src/Window.cpp [hunk 1]
```diff
@@ -1090,7 +1090,8 @@ void Window::dispatchCommand(QString command) {
 QString Window::executeCommand(QString command) {
     QStringList tokens = command.split(" ", QString::SkipEmptyParts);
     if (tokens.size() != 1) {
-        return INVALID;
+        if (! (tokens.size() == 2 && tokens.at(0) == "moveForward" ))
+            return INVALID;
     }
```
Comment (mackorone) [inline src/Window.cpp:1095]: I think it's slightly cleaner to check the easy case and the harder case separately, so that you can use the `function` variable in the latter, rather than `tokens.at(0)`. Also, my preference is to *always* use braces, just to be safe.
```cpp
if (tokens.size() < 1 || tokens.size() > 2) {
  return INVALID;
}
QString function = tokens.at(0);
if (tokens.size() == 2 && function != "moveForward") {
    return INVALID;
}
```

File: src/Window.cpp [hunk 2]
```diff
@@ -1109,7 +1110,14 @@ QString Window::executeCommand(QString command) {
...
     else if (function == "moveForward") {
-        bool success = moveForward();
+        bool success;
+        if (tokens.size() == 2) {
+            int distance = tokens.at(1).toInt();
+            success = moveForward(distance);
+        }
+        else {
+            success = moveForward();
+        }
```
Comment (mackorone) [inline src/Window.cpp:1120]: Rather than optionally providing a distance to the `moveForward()` argument, just use a default value of `1` if a second argument isn't provided. Then we don't need a default argument in the `moveForward()` function:

```cpp
int distance = 1;
if (tokens.size() == 2) {
    distance = tokens.at(1).toInt();
}
bool success = moveForward(distance);

File: src/Window.h [hunk 3]
```diff
@@ -168,23 +168,27 @@ class Window : public QMainWindow {
...
     QSlider* m_speedSlider;
 
     double progressRequired(Movement movement);
     void updateMouseProgress(double progress);
     void scheduleMouseProgressUpdate();
     bool isMoving();
 
+    // ----- Scoreboard -----
+    Stats* stats;
+
```
Comment (mackorone) [inline src/Window.h:181]: This looks like it belongs in a different commit. Mind removing for now?

File: src/Window.cpp [hunk 4]
```diff
@@ -1299,8 +1339,10 @@ bool Window::wallLeft() {
     return isWall({position.first, position.second, direction});
 }
 
-bool Window::moveForward() {
-    if (wallFront()) {
+bool Window::moveForward(int distance) {
+    moveLength = distance;
+    // check for crashes and negative distances
+    if (distance < 0 || wallFront(true)) { // the true parameter tells wallFront() to look ahead for moveLength tiles
```
Comment (mackorone) [inline src/Window.cpp:1345]: As written, the simulator will immediately respond `crash` if the mouse can't move forward the request number of spaces. I wonder... would it be crazy for the mouse to move forward as many spaces as possible *and then* crash? That'd be a tad more realistic, right? If that sounds reasonable to you, you should store `m_forwardMovesRemaining` and decrement by one after each successful movement forward. If you hit zero, you can return `ACK`. If you encounter a wall before hitting zero, you can return `CRASH`.

File: src/Window.cpp [hunk 5]
```diff
@@ -1279,9 +1287,41 @@ int Window::mazeHeight() {
     return m_maze->getHeight();
 }
 
-bool Window::wallFront() {
+bool Window::wallFront(bool useDistance) {
```
Comment (mackorone) [inline src/Window.cpp:1290]: My guess is that we don't actually need to make any changes to this function, assuming that we move forward as far as possible as described in the comment below.

File: README.md [hunk 6]
```diff
@@ -91,6 +91,7 @@ bool wallRight();
 bool wallLeft();
 
 void moveForward();  // can result in "crash"
+void moveForward(int distance); // can result in "crash"
```
Comment (mackorone) [inline README.md:94]: Might as well use the default argument syntax here, rather than method overloading:

```cpp
void moveForward(int distance = 1); // can result in "crash"
```

File: README.md [hunk 7]
```diff
@@ -134,9 +135,10 @@ void ackReset();
 * **Action:** None
 * **Response:** `true` if there is a wall to the left of the robot, else `false`
 
-#### `moveForward`
-* **Args:** None
-* **Action:** Move the robot forward by one cell
+#### `moveForward distance`
+* **Args:**
+  * `distance` - (optional) The number of cells to move
+* **Action:** Move the robot forward by the specified number of cells, 1 cell by default
```
Comment (mackorone) [inline README.md:141]: Nitpick: spell out "one"

File: src/Window.cpp (hunk 4 above)
Comment (zdasaro) [reply src/Window.cpp:1345]: I agree that this is more realistic. However, I'm not sure how/where to implement this. Currently, the way `processQueuedCommands()` is written, the mouse cannot move at all after a command that results in "crash".

File: src/Window.h [hunk 8]
```diff
@@ -180,11 +181,11 @@ class Window : public QMainWindow {
     int mazeWidth();
     int mazeHeight();
 
-    bool wallFront();
+    bool wallFront(bool useDistance = false);
```
Comment (mackorone) [inline src/Window.h:184]: Can you change this to `int distance`? My preference is to avoid default values for arguments, which means you'll need to explicit pass `1` when the user calls `wallFront`.

File: src/Window.h [hunk 9]
```diff
@@ -168,6 +168,7 @@ class Window : public QMainWindow {
     Movement m_movement;
     double m_movementProgress;
     double m_movementStepSize;
+    int moveLength;
```
Comment (mackorone) [inline src/Window.h:171]: I think you can just remove this

File: src/Window.cpp [hunk 10]
```diff
@@ -1299,11 +1338,14 @@ bool Window::wallLeft() {
...
-bool Window::moveForward() {
-    if (wallFront()) {
+bool Window::moveForward(int distance) {
+    moveLength = distance;
+    // check for crashes and negative distances
+    if (distance < 0 || wallFront(true)) { // the true parameter tells wallFront() to look ahead for moveLength tiles
         return false;
     }
     m_movement = Movement::MOVE_FORWARD;
+    scheduleMouseProgressUpdate();
```
Comment (mackorone) [inline src/Window.cpp:1348]: This shouldn't be necessary

File: src/Window.cpp (hunk 4 above)
Comment (mackorone) [reply src/Window.cpp:1344]: Yea, good point, it'd be a bit more complicated. I think it's fine to leave the logic as is for now.

However, rather than storing `distance` in an instance variable and then passing `true` to `wallFront`, how about just passing `distance` into `wallFront`?

**Comment by mackorone** [comment]
@zdasaro sorry, in case it wasn't clear: please make the change I mentioned above and then I'll merge this in.

File: src/Window.cpp (hunk 4 above)
Comment (zdasaro) [reply src/Window.cpp]: I'll make the change right now to pass `distance` to wallFront. However, the instance variable `moveLength` is still in use in the function `updateMouseProgress()`. Is that alright?

File: src/Window.cpp (hunk 10 above)
Comment (zdasaro) [reply src/Window.cpp:1346]: Sorry, didn't mean to add that.

**Comment by zdasaro** [comment]
I think I've made the changes you requested, but `moveLength` is still a topic of discussion.

File: src/Window.cpp (hunk 4 above)
Comment (mackorone) [reply src/Window.cpp]: Oh, right, totally forgot. Yea, that's fine :).
